Add plaid link action and controller

vinothvino42 opened the pull request:
In this PR, I added the plaid link action

**Sample YAML**
```yaml
Button:
  label: Link Account
  onTap:
    openPlaidLink:
      id: plaidId
      linkToken: YOUR LINK TOKEN
      onSuccess: |
        //@code
        console.log("Account Linked Successfully");
        console.log(event.data);
        console.log("Public Token" + event.data['publicToken']);

      onEvent: |
        //@code
        console.log("All Events");
        console.log(event.data);

      onExit: |
        //@code
        console.log("Something went wrong!");
        console.log(event.data);
```

`event.data` will be different for each callback. 
onSuccess - [linkSuccess object](https://plaid.com/docs/link/ios/#link-ios-onsuccess-linkSuccess)
onEvent - [linkEvent object](https://plaid.com/docs/link/ios/#link-ios-onevent-linkEvent)
onExit - [linkExit object](https://plaid.com/docs/link/ios/#link-ios-onexit-linkExit)

Please check out the doc for more info - https://plaid.com/docs/link/ios/#onsuccess

vusters commented on lib/screen_controller.dart:
```diff
@@ -185,6 +188,51 @@ class ScreenController {
+                scopeManager!,
+                action.onEvent!,
+                event: EnsembleEvent(action.initiator!,
+                    data: {'eventData': linkEvent.toJson()}),
```
by convention we always do `event.data.*`. So do `data: linkEvent.toJson()`. Same with others

vinothvino42 replied:
We have three callbacks with data, so I thought to split it evenly with their name like eventData, successData, errorData. Okay I'll add every object directly to the data param.

vinothvino42 replied:
Done

vusters commented on lib/widget/plaid_link/plaid_link_controller.dart:
```diff
@@ -0,0 +1,36 @@
+import 'package:flutter/material.dart';
+import 'package:plaid_flutter/plaid_flutter.dart';
+
+typedef PaidLinkSuccessCallback = void Function(LinkSuccess);
```
spelling

vinothvino42 replied:
Okay I'll fix it

vinothvino42 replied:
Done

vinothvino42 commented on lib/framework/action.dart:
```diff
@@ -211,6 +211,39 @@ class ShowBottomModalAction extends EnsembleAction {
+
+    return PlaidLinkAction(
+      initiator: initiator,
+      options: Utils.getMap(payload['options']),
```
Okay

vusters commented:
One more thing. Please create a new Module, call it `ensemble_connect` and put the Plaid code in here. The base ensemble shouldn't have any reference to Plaid. See existing modules like 'ensemble_auth' for example.

vinothvino42 commented:
@vusters @kmahmood74 Please see the above updated YAML EDL. 
@vusters I updated the changes which you've mentioned here as a comment and I just moved the paid link service to a separate package - https://github.com/EnsembleUI/ensemble_connect

You can enable/disable the paid link service as mentioned here - https://github.com/EnsembleUI/starter#modules
```dart
if (usePlaidLink) {
      // Uncomment to enable plaid link service
      GetIt.I.registerSingleton<PlaidLinkManager>(PlaidLinkManagerImpl());
    } else {
      GetIt.I.registerSingleton<PlaidLinkManager>(PlaidLinkManagerStub());
    }
```

vusters commented:
Ensemble_Live requires all module to turn on. Please enable Connect too - see existing example.

vinothvino42 commented:
> Ensemble_Live requires all module to turn on. Please enable Connect too - see existing example.

I'll add that once this PR merged it. Because in the preview it supposed to map to main branch, this one is in plaid-integration branch
